data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3Source.java

```diff
@@ -41,18 +43,25 @@ public class S3Source implements Source<Record<Event>>, UsesSourceCoordination {
     private final PluginFactory pluginFactory;
     private final Optional<S3ScanScanOptions> s3ScanScanOptional;
     private final AcknowledgementSetManager acknowledgementSetManager;
+    private final AwsCredentialsSupplier awsCredentialsSupplier;
     private final boolean acknowledgementsEnabled;
     private SourceCoordinator<S3SourceProgressState> sourceCoordinator;
 
 
     @DataPrepperPluginConstructor
-    public S3Source(PluginMetrics pluginMetrics, final S3SourceConfig s3SourceConfig, final PluginFactory pluginFactory, final AcknowledgementSetManager acknowledgementSetManager) {
+    public S3Source(
+            final PluginMetrics pluginMetrics,
+            final S3SourceConfig s3SourceConfig,
+            final PluginFactory pluginFactory,
+            final AcknowledgementSetManager acknowledgementSetManager,
+            final AwsCredentialsSupplier awsCredentialsSupplier) {
```

Comment:
QUES: how does this get injected?

Reply:
When registering an extension, Data Prepper registers a prototype bean to an application context shared by all plugins. You can see the code here:

https://github.com/opensearch-project/data-prepper/blob/3ce5b53507312d06fe2edc90e64fe8224b357978/data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/DataPrepperExtensionPoints.java#L30-L36

Later, when this constructor is used, the plugin framework requests this from the application context. You can see this in the next bit of code.

https://github.com/opensearch-project/data-prepper/blob/3ce5b53507312d06fe2edc90e64fe8224b357978/data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/ComponentPluginArgumentsContext.java#L102

When that happens, Data Prepper will call `provideInstance`. This goes into the AWS Plugin:

https://github.com/opensearch-project/data-prepper/blob/565538323d80643d87781a3d8149279ee5f9dda4/data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsExtensionProvider.java#L20-L23